### PR TITLE
Adds struct/contract

### PR DESCRIPTION
**Pull request by tommymchugh**
Implementation for one of the features requested in #1699

### CODE REVIEW

**Comment by rfindler** [comment]
Thanks for your interest and work!

I am sorry, however, it is hard to see what you have changed here, due to
the very large number of changes to white space. Can you please separate
those out?

Robby

**Comment by rfindler** [comment]
Oh, sorry! I have found the "ignore whitespace changes" feature in the github GUI now. I'll take a look.

**Comment by rfindler** [comment]
I think the change has to be a little more nuanced because `struct` doesn't bind the constructor as `make-...`, the "make-" prefix isn't there.

Also, it looks like you've moved ever so slightly away from the style guide guidelines by moving a helper function into a `let*` out of a `define`.

**Comment by tommymchugh** [comment]
Gotcha! Thanks for the direction on this, I'll work on this

**Comment by tommymchugh** [comment]
@rfindler Sorry for the delay getting back on this, was on holiday past few days. This now takes into account all the differences between struct and define-struct.

**Comment by tommymchugh** [comment]
43e4128 cleans code to remove whitespaces that were very annoying to review code with.

File: racket/collects/racket/contract/region.rkt [hunk 1]
```diff
@@ -123,332 +124,392 @@
...
+                                                            (list
+                                                             #'(define-syntax struct-name
+                                                                 (make-struct-info
+                                                                  (λ ()
+                                                                    (list #f #f #f
+                                                                          (list #f) (list #f)
+                                                                          #,(quoter #'super))))))]))]
+                                                      [define-returned-value-temps (if (equal? type 'define-struct/contract)
+                                                                                       #`(define maker-value ... temp-maker-name ...)
+                                                                                       (void))])
```
Comment (rfindler) [inline racket/collects/racket/contract/region.rkt]: This should be #`(void), not (void).

**Comment by rfindler** [comment]
It looks like this part is basically ready. The docs update is missing. I can do that if you'd like.

Also, I see that the program below produces the unfortunate output below it (there is an extra argument showing in the contract in the error message), but this doesn't seem to be a problem with the changes you're doing here, as it also happens for `define-struct/contract`.

```
#lang racket
(struct/contract s ([x integer?][y boolean?]))
(s 1 2)
```

produces this output:

```
$  racket ~/tmp.rkt
s: contract violation
  expected: boolean?
  given: 2
  in: the 2nd argument of
      (-> integer? boolean? symbol? any)
  contract from: (struct s)
  blaming: /Users/robby/tmp.rkt
   (assuming the contract is correct)
```

**Comment by tommymchugh** [comment]
@rfindler I took care of the docs and the syntax quasiquote addition, and updated the define-struct/contract to better match the styling of struct and define-struct relationship in the docs. 

Yeah, in regards to the weird error statement, without looking at it deeply, I think it has to be because of the structure of the `#:guard` procedure that was used to enforce the contract, which isn't changed with these additions. I'll file an issue and take a little deeper look over next few days to see what's going on with that and see if we can change anything to better reflect the contract violation

**Comment by rfindler** [comment]
Okay, can you squash and rebase and then I'll merge?

**Comment by tommymchugh** [comment]
Done. Once tests pass again, should be good to go.

**Comment by rfindler** [comment]
Thanks!

**Pull request merged**